Remotion/Core/Development.Analyzers/RMCORE0001_SafeContextAnalyzer.cs

```diff
@@ -0,0 +1,277 @@
+        if (executionSymbol != null)
+        {
+          executionRunMethodSymbol = GetMethodSymbol(executionSymbol, "Run");
+
+        }
+
+        IReadOnlyCollection<IMethodSymbol>? parallelForMethodSymbols = null;
+        IReadOnlyCollection<IMethodSymbol>? parallelForEachMethodSymbols = null;
+        IReadOnlyCollection<IMethodSymbol>? parallelForEachAsyncMethodSymbols = null;
+        IReadOnlyCollection<IMethodSymbol>? parallelInvokeMethodSymbols = null;
+
+        if (parallelSymbol != null)
+        {
+          parallelForMethodSymbols = GetMethodSymbols(parallelSymbol, "For");
+          parallelForEachMethodSymbols = GetMethodSymbols(parallelSymbol, "ForEach");
+          parallelForEachAsyncMethodSymbols = GetMethodSymbols(parallelSymbol, "ForEachAsync");
+          parallelInvokeMethodSymbols = GetMethodSymbols(parallelSymbol, "Invoke");
+        }
+
+        IReadOnlyCollection<IMethodSymbol>? ctorThreadingThreadMethodSymbols = null;
+
+        if (threadSymbol != null)
+        {
+          ctorThreadingThreadMethodSymbols = GetCtorSymbols(threadSymbol);
+        }
+
+        IReadOnlyCollection<IMethodSymbol>? ctorThreadingTimerMethodSymbols = null;
+        if (threadingTimerSymbol != null)
+        {
+          ctorThreadingTimerMethodSymbols = GetCtorSymbols(threadingTimerSymbol);
+        }
+
+
+        var symbolContext = new MethodSymbolContext(
+            taskRunMethodSymbols,
+            taskContinueWithMethodSymbol,
+            parallelForMethodSymbols,
+            parallelForEachMethodSymbols,
+            parallelForEachAsyncMethodSymbols,
+            parallelInvokeMethodSymbols,
+            ctorThreadingThreadMethodSymbols,
+            ctorThreadingTimerMethodSymbols,
+            executionRunMethodSymbol);
+
+
+        compilationContext.RegisterSyntaxNodeAction(
+            ctx => AnalyzeInvocationExpression(ctx, symbolContext),
+            SyntaxKind.InvocationExpression);
+
+        compilationContext.RegisterSyntaxNodeAction(ctx => AnalyzeObjectCreationExpression(ctx, symbolContext, new ErrorMessageParameter()), SyntaxKind.ObjectCreationExpression);
+      });
+    }
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(AlternativeDescriptor, WarningDescriptor);
+
+    private static void AnalyzeInvocationExpression (SyntaxNodeAnalysisContext ctx, MethodSymbolContext methodSymbols)
+    {
+      IMethodSymbol outMethodSymbol;
+      var errorMessageParameter = new ErrorMessageParameter();
+      var invocationSymbolInfo = ctx.SemanticModel.GetSymbolInfo(ctx.Node);
+
+      if (invocationSymbolInfo.Symbol == null)
+        return;
+
+      var invocationOriginalDefinitionSymbol = invocationSymbolInfo.Symbol.OriginalDefinition;
+
+      if (methodSymbols.TaskRunMethodSymbols != null && TryFindItem(methodSymbols.TaskRunMethodSymbols, s => SymbolEqualityComparer.Default.Equals(invocationOriginalDefinitionSymbol, s), out outMethodSymbol))
```

Comment:
We are repeating a lot of the same diagnostic generation logic. We could change `MethodSymbolContext` to contain a list of `(IMethodSymbol method, Action<SyntaxNodeAnalysisContext , IMethodSymbol>)` and have all the "what error has what message" logic be moved into the `Initialize` method. This would simplify the logic here and probably also in `Initialize`. What do you think?